ConcurrentAPI.adoc

```diff
@@ -0,0 +1,244 @@
+== Distributed Concurrent API
+
+=== Motivation
+
+Coming from an initial issue where we need to provide an implementation for vert.x cluster manager for the shared data feature,
+provide concurrent and distributed programming capabilities.
+
+* Distributed Locks
+* Distributed CountDownLatch
+
+=== Scope
+
+All these interfaces should be placed in commons, in order to be used either in Embedded and Server mode.
+
+== Infinispan Distributed Locks
+
+=== DistributedLockManager and configuration
+
+```java
+package org.infinispan.lock.api;
+
+public class DistributedLockManager {
+   boolean defineLock(LockConfiguration configuration);
+   InfinispanLock get(String name);
+   LockConfiguration getConfiguration(String name);
+   boolean isDefined(String name);
+   CompletableFuture<Boolean> remove(String name);
+}
+
+
+public class LockConfiguration {
+   private final String name; // default "ISPN-Lock"
+   private final boolean rentrant; // default true
```

Comment:
@rvansa vert.x needs a non rentrant lock for their SharedData functionality. Internally, vert-x hazelcast cluster manager uses a semaphore with a single permit to achieve this. At first I was about to propose that we implement distributed semaphores. But after a discussion with Will, we decided to offer a Lock. This lock can handle different levels of rentrancy. So for vert-x usage, would behave as a semaphore with a single permit, but we can offer our users clustered locks rentrant per thread, or rentrant per cluster node for example. And we can start with a non rentrant implementation that will be used by vert-x/infinispan cluster manager.